Feature/pre processing

ciaradunks opened the pull request:
Pull request including:

- [ ] Updated pre-processing to include the option for adding custom attributes e.g. renewable factor, emission factor, land requirement
- [ ] Included general_custom_attributes example to show how the user can input them and how they get added to the CSV file
- [ ] Included tests for pre-processing of custom attributes
- [ ] Included layout for multi-objective optimization (MOO) - not completed yet

Commit:
Include custom attributes in pre-processing

Commit:
Include custom attributes in pre-processing

Commit:
Adapt compute.py file

Commit:
First test for pre-processing custom attributes

Commit:
First test for pre-processing custom attributes

Commit:
Additions to custom attributes testing

Commit:
Additions to custom attributes testing

Commit:
New example for using custom attributes

Commit:
New example for using custom attributes

Commit:
Updates to pre_processing including layout for MOO

Commit:
Updates to pre_processing including layout for MOO

Commit:
Updated tests for pre-processing

Commit:
Updated tests for pre-processing

Commit:
Updated compute.py file

Bachibouzouk committed:
Adapt compute.py file

Bachibouzouk committed:
Updated compute.py file

Bachibouzouk committed:
Fix failing tests

Bachibouzouk commented on src/oemof_tabular_plugins/general/pre_processing_moo.py:
```diff
@@ -0,0 +1,34 @@
+def pre_processing_moo():
```
DocumentationDrivenDevelopment, very nice :)

Bachibouzouk commented on tests/test_pre_processing.py:
```diff
@@ -347,4 +347,162 @@ def test_no_annuity_no_cost_params_logs_message(self, caplog):
+        csv file contains custom attributes and they have been defined as list by user.
+        """
+        # copy scenario csv file and datapackage json file to the package path
+        f_name = "cust_attr.csv"
+        dp_name = "dp_no_output_params.json"
+        shutil.copy(os.path.join(self.test_inputs_pre_p, f_name), os.path.join(self.package_path, f_name))
+        shutil.copy(os.path.join(self.test_inputs_pre_p, dp_name),
+                    os.path.join(self.pre_p_dir, "datapackage.json"))
```
This code is redundant, I would move it within a method which takes f_name and dp_name as arguments

Bachibouzouk commented on src/oemof_tabular_plugins/general/pre_processing.py:
```diff
@@ -2,227 +2,341 @@
+    # if custom attributes were found, update the datapackage.json file
+    if has_custom_attributes:
+        update_datapackage_json_custom_attributes(scenario_dir, element)
+    return
```
It returns None, not what the docstring mentions. I would rather remove the mention in the return docstring description

Bachibouzouk commented on src/oemof_tabular_plugins/general/pre_processing.py:
```diff
@@ -2,227 +2,341 @@
+    :param wacc: weighted average cost of capital (WACC) applied throughout the model (%)
+    :param custom_attributes: list of custom attributes included in the model (defined in compute.py), default is None
+    :param moo: whether the multi-objective optimization is activated, default is False
+    :return: updated input scenario data
```
Same comment as above regarding the return value

Bachibouzouk commented on src/oemof_tabular_plugins/general/pre_processing_moo.py:
```diff
@@ -0,0 +1,34 @@
+
+    The outcome is that the main costs 'capacity_cost' will be replaced by an aggregated
+    indicator representing the multi-objective optimization goals.
+    This function will replace the pre_processing_costs function if moo is set to True
```
it will also replace `pre_processing_custom_attributes` right?

Bachibouzouk commented on src/oemof_tabular_plugins/general/pre_processing_moo.py:
```diff
@@ -0,0 +1,34 @@
+        -> the value is normalised based on e.g. global availability
+    Then these values are added together
+
+    User includes specific cost, specific emission factor, specific land requirement, specific water footprint
```
Here the exact mathematical calculation could be written down explicitly. 

Dummy arguments could be provided to match the arguments description in the docstring

ciaradunks replied:
I have done the dummy arguments, but I the exact mathematical calculation is yet to be determined (Julian's task)

Bachibouzouk reviewed (changes requested):
Nice work @ciaradunks , a lot of news tests and good docstrings :) 

I made a few comments within the code

One thing I still see missing is the change in the main README to explain how the user can run those example scripts. Doesn't need to be super complex, but for example it should mention that one has to run `python example/compute.py`

Commit:
Fixed merge conflict

Commit:
Changes from review comments

ciaradunks commented:
Thanks for the helpful review @Bachibouzouk ! I have addressed most of your comments! I totally agree with including an explanation in the README file. I will do this soon in a separate PR when I focus on documentation

Bachibouzouk commented on examples/scripts/compute.py:
```diff
@@ -1,5 +1,8 @@
 import os
+<<<<<<< HEAD
```
this is not fixed, you committed the conflict ...
